### PR TITLE
Merge critical fixes from Cadence

### PR DESCRIPTION
**Pull request by alexshtin**
https://github.com/uber/cadence/pull/3376
https://github.com/uber/cadence/pull/3458

### CODE REVIEW

File: service/worker/replicator/replicator.go [hunk 1]
```diff
@@ -165,19 +168,32 @@ func (r *Replicator) createKafkaProcessors(currentClusterName string, clusterNam
...
 		r.clientBean.GetRemoteAdminClient(clusterName),
 		common.CreateAdminServiceRetryPolicy(),
-		common.IsWhitelistServiceTransientError,
+		isRetryableError,
+	)
+	// Create retry policy for service busy error
+	adminRetryClient := admin.NewRetryableClient(
+		adminClient,
+		common.CreateReplicationServiceBusyRetryPolicy(),
+		common.IsResourceExhausted,
```
Comment (wxing1292) [inline service/worker/replicator/replicator.go:177]: `IsResourceExhausted` already contains `ResourceExhausted`
plz remove this logic

File: service/worker/replicator/replicator.go [hunk 2]
```diff
@@ -165,19 +168,32 @@ func (r *Replicator) createKafkaProcessors(currentClusterName string, clusterNam
...
-		common.IsWhitelistServiceTransientError,
+		isRetryableError,
 	)
+	// Create retry policy for service busy error
+	historyRetryClient := history.NewRetryableClient(
+		historyClient,
+		common.CreateReplicationServiceBusyRetryPolicy(),
+		common.IsResourceExhausted,
+	)
+
```
Comment (wxing1292) [inline service/worker/replicator/replicator.go:190]: `IsResourceExhausted` already contains `ResourceExhausted`
plz remove this logic

File: service/history/historyEngine.go [hunk 3]
```diff
@@ -293,11 +294,26 @@ func NewEngineWithShardContext(
...
+			common.IsResourceExhausted,
+		)
 		nDCHistoryResender := xdc.NewNDCHistoryResender(
 			shard.GetNamespaceCache(),
-			shard.GetService().GetClientBean().GetRemoteAdminClient(currentClusterName),
+			adminRetryableClient,
 			func(ctx context.Context, request *historyservice.ReplicateEventsV2Request) error {
-				return historyEngImpl.ReplicateEventsV2(ctx, request)
+				_, err := historyRetryableClient.ReplicateEventsV2(ctx, request)
+				return err
```
Comment (wxing1292) [inline service/history/historyEngine.go:316]: here, the logic should target directly at the local shard (this shard).

my guess, the idea here is to wrap the ReplicateEventsV2 API call (i.e. historyRetryableClient) for retry
the only easy change is to use the retryable client, but the solution involves network stack (less efficiency)

let us cut a ticket to unify the client / server interface and update the logic here

File: service/history/replicationTaskFetcher.go [hunk 4]
```diff
@@ -284,7 +299,9 @@ func (f *ReplicationTaskFetcherImpl) getMessages(
 	}
 	response, err := f.remotePeer.GetReplicationMessages(ctx, request)
 	if err != nil {
-		return nil, err
+		if _, ok := err.(*serviceerror.ResourceExhausted); !ok {
+			return nil, err
+		}
 	}
 
 	return response.GetShardMessages(), err
```
Comment (wxing1292) [inline service/history/replicationTaskFetcher.go:307]: this logic is weird, i.e.
if err is not nil and of type `ResourceExhausted` -> `response.GetShardMessages()` causing NPE?

i guess the previous logic is good for now
```go
if err != nil {
	return nil, err
}
```

Comment (wxing1292) [reply service/history/replicationTaskFetcher.go:307]: actually, turns out the proto layer already handles the nil case.

i think it is still preferable to do the following```go
if err != nil {
	return nil, err
}
```
to make the code less confusing

**Review by wxing1292 — APPROVED** [review]
Mostly looks ok

File: service/history/replicationTaskFetcher.go (hunk 4 above)
Comment (alexshtin) [reply service/history/replicationTaskFetcher.go]: I don't see where it handles `nil` case. If `err` is not `nil` response is always `nil`. I am removing this change.

File: service/worker/replicator/replicator.go (hunk 1 above)
Comment (alexshtin) [reply service/worker/replicator/replicator.go:177]: I don't understand this comment. Retry policy is different.

Comment (alexshtin) [reply service/worker/replicator/replicator.go:177]: I removed `ResourceExhausted` from `isRetryableError` though.

**Pull request merged**